sdks/java/core/src/main/java/org/apache/beam/sdk/util/OrderedCode.java

```diff
@@ -0,0 +1,764 @@
+   * byte array, followed by a separator and appends the result to its
+   * internal encoded byte array store.
+   *
+   * <p>It works with the input array,
+   * so the input array 'value' should not be modified till the method returns.
+   *
+   * @param value bytes to be written.
+   * @see #readBytes()
+   */
+  public void writeBytes(byte[] value) {
+    writeBytes(value, false);
+  }
+
+  public void writeBytesDecreasing(byte[] value) {
+    writeBytes(value, true);
+  }
+
+  private void writeBytes(byte[] value, boolean invert) {
+    // Determine the length of the encoded array
+    int encodedLength = 2;      // for separator
+    for (byte b : value) {
+      if ((b == ESCAPE1) || (b == ESCAPE2)) {
+        encodedLength += 2;
+      } else {
+        encodedLength++;
+      }
+    }
+
+    byte[] encodedArray = new byte[encodedLength];
+    int copyStart = 0;
+    int outIndex = 0;
+    for (int i = 0; i < value.length; i++) {
+      byte b = value[i];
+      if (b == ESCAPE1) {
+        arraycopy(invert, value, copyStart, encodedArray, outIndex,
+            i - copyStart);
+        outIndex += i - copyStart;
+        encodedArray[outIndex++] = convert(invert, ESCAPE1);
+        encodedArray[outIndex++] = convert(invert, NULL_CHARACTER);
+        copyStart = i + 1;
+      } else if (b == ESCAPE2) {
+        arraycopy(invert, value, copyStart, encodedArray, outIndex,
+            i - copyStart);
+        outIndex += i - copyStart;
+        encodedArray[outIndex++] = convert(invert, ESCAPE2);
+        encodedArray[outIndex++] = convert(invert, FF_CHARACTER);
+        copyStart = i + 1;
+      }
+    }
+    if (copyStart < value.length) {
+      arraycopy(invert, value, copyStart, encodedArray, outIndex,
+          value.length - copyStart);
+      outIndex += value.length - copyStart;
+    }
+    encodedArray[outIndex++] = convert(invert, ESCAPE1);
+    encodedArray[outIndex] = convert(invert, SEPARATOR);
+
+    encodedArrays.add(encodedArray);
+  }
+
+  private static byte convert(boolean invert, byte val) {
+    return invert ? (byte) ~val : val;
+  }
+
+  private static byte invert(byte val) {
+    return convert(true, val);
+  }
+
+  private void arraycopy(
+      boolean invert, byte[] src, int srcPos, byte[] dest, int destPos, int length) {
+    System.arraycopy(src, srcPos, dest, destPos, length);
+    if (invert) {
+      for (int i = destPos; i < destPos + length; i++) {
+        dest[i] = (byte) ~dest[i];
+      }
+    }
+  }
+
+  /**
+   * Encodes the long item, in big-endian format, and appends the result to its
+   * internal encoded byte array store.
+   *
+   * @see #readNumIncreasing()
+   */
+  public void writeNumIncreasing(long value) {
+    // Values are encoded with a single byte length prefix, followed
+    // by the actual value in big-endian format with leading 0 bytes
+    // dropped.
+    byte[] bufer = new byte[9];  // 8 bytes for value plus one byte for length
```

Comment:
You can use this [Long#numberOfLeadingZeros](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#numberOfLeadingZeros-long-) to find out how many bytes you will need to encode this long. Then  you can just allocate that many bytes instead of allocating two arrays.

Reply:
will need to fix underlying library